File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <climits>

using namespace std;

// start, end, cost
int solution(int n, int s, int a, int b, vector<vector<int>> fares) {
    int answer = INT_MAX;
    
    // 모든 노드에서 다른 모든 노드로 가는 최단 거리 파악
    int graph[n+1][n+1];
    for(auto& g:graph){
        fill(g, g+n+1, 20000001);
    }
    for(int i = 1; i <= n; i++){
        graph[i][i] = 0;
    }
    for(auto&f: fares){
        int st = f[0]; int end = f[1]; int cost = f[2]; 
        graph[st][end] = cost; graph[end][st] = cost;
    }
    
    // 플로이드워셜로 모든 노드에서 모든 노드로 가는 최단 거리 계산
    for(int i = 1; i <= n; i++){ // 거쳐가는 노드
        for(int start = 1; start <= n; start++){
            for(int end = 1; end <= n; end++){
                graph[start][end] = 
                    min(graph[start][end], graph[start][i] + graph[i][end]);
            }
        }
    }
    
    // a와 b에서 1~n까지의 중간 지점 정해놓고 비용 판단
    for(int i = 0; i <= n; i++){
        answer = min(answer, graph[s][i] + graph[i][a] + graph[i][b]); 
    }
    // 그냥 s -> a, s -> b
    answer = min(answer, graph[s][a]+graph[s][b]);
    
    // 최저 예상 택시요금을 계산해서 return
    return answer;
}

File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.java
import java.util.*;

class Node{
    int to; int cost; 
    Node(int to, int cost){
        this.to = to; 
        this.cost = cost;
    }
}
class Solution {
    
    int MAX_INT = Integer.MAX_VALUE;
    
    public int solution(int n, int s, int a, int b, int[][] fares) { // [c, d, f] 
        int answer = 0;
        
        int[][] dist = new int[n+1][n+1]; 
        for(int i = 1; i <= n; i++){
            Arrays.fill(dist[i], MAX_INT);
            dist[i][i] = 0;
        }

        for(int[] f : fares){
            int c = f[0]; int d = f[1]; int cost = f[2]; 
            dist[c][d] = cost; 
            dist[d][c] = cost; 
        }
        
        // 각 노드에서 각 노드로의 최소 비용 구하기 -> 플로이드 워셜
        for(int i = 1; i <= n; i++){ // 중간
            for(int j = 1; j <= n; j++){ // 시작
                for(int k = j+1; k <= n; k++){ // 끝
                    // 시작-끝 보다 시작-중간-끝이 더 짧으면 갱신
                    if(dist[j][i] == MAX_INT || dist[i][k] == MAX_INT) continue;  
                    if(dist[j][i] + dist[i][k] < dist[j][k]){
                        dist[j][k] = dist[j][i] + dist[i][k];
                        dist[k][j] = dist[j][i] + dist[i][k];
                    }
                    
                }
            }
        }
        
        // s부터 a, b까지의 최단 거리 구하기
        answer = dist[s][a] + dist[s][b];
        
        // s부터 중간 지점 -> a, b 최단 거리 구하기
        for(int i = 1; i <= n; i++){ // 중간 지점
            if(dist[s][i] == MAX_INT || dist[i][a] == MAX_INT || dist[i][b] == MAX_INT) continue;
            if(dist[s][i] + dist[i][a] + dist[i][b] < answer) {
                answer = dist[s][i] + dist[i][a] + dist[i][b];
            }
        }
        
        
        return answer; // 최저 예상 택시요금
    }
    
}